src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -301,6 +302,65 @@ protected void ShareTemps(ImmutableArray<BoundDecisionDagNode> sortedNodes)
                         }
                     }
                 }
+
+                if (loweredInput.Kind == BoundKind.TupleLiteral &&
+                    !decisionDag.TopologicallySortedNodes.Any(n => !usesOriginalInput(n)) &&
+                    false)
+                {
+                    // If the switch governing expression is a tuple literal that is not used anywhere,
+                    // (though perhaps its component parts are used), then we can save the component parts
+                    // and assign them into temps (or perhaps user variables) to avoid the creation of
+                    // the tuple altogether.
+                    decisionDag = RewriteTupleSwitch(decisionDag, (BoundTupleLiteral)loweredInput, addCode);
+                }
+                else
+                {
+                    // Otherwise we emit an assignment of the input expression to a temporary variable.
+                    BoundExpression inputTemp = _tempAllocator.GetTemp(inputDagTemp);
+                    if (inputTemp != loweredInput)
+                    {
+                        addCode(_factory.AssignmentExpression(inputTemp, loweredInput));
+                    }
+                }
+
+                return decisionDag;
+
+                bool usesOriginalInput(BoundDecisionDagNode node)
+                {
+                    switch (node)
+                    {
+                        case BoundWhenDecisionDagNode n:
+                            return (n.Bindings.Any(b => b.dagTemp.IsOriginalInput));
+                        case BoundTestDecisionDagNode t:
+                            return t.Test.Input.IsOriginalInput;
+                        case BoundEvaluationDecisionDagNode e:
+                            switch (e.Evaluation)
+                            {
+                                case BoundDagFieldEvaluation f:
+                                    // Uses only a tuple element, not the whole input
+                                    return f.Input.IsOriginalInput && !f.Field.IsTupleElement();
+                                default:
+                                    return e.Evaluation.Input.IsOriginalInput;
```

Comment:
>e.Evaluation.Input.IsOriginalInput [](start = 43, length = 34)

Minor point: Consider checking this common expression outside of the `switch (e.Evaluation) { ... }`.

Reply:
The value being switched on does not have an `Input` field, and `e` is not in scope before the switch.

---
In reply to: [179661343](https://github.com/dotnet/roslyn/pull/25963#discussion_r179661343) [](ancestors = 179661343)

Reply:
I simply meant check `e.Evaluation.Input.IsOriginalInput` before the nested `switch`, but it was a minor point.

Reply:
I understand now. I think this is more clear and concise (I would have to add a new if statement to return false when)

---
In reply to: [179859796](https://github.com/dotnet/roslyn/pull/25963#discussion_r179859796) [](ancestors = 179859796)